pkg/cli/options.go

```diff
@@ -139,3 +150,150 @@ func WithCompletion() Option {
+		}
+
+		pluginsRoot = filepath.Join(userHomeDir, ".config", "kubebuilder", "plugins")
+
+	default:
+		// freebsd, openbsd, windows...
+		logrus.Errorf("Host not supported: %s", host)
+	}
+
+	return pluginsRoot
+}
+
+// DiscoverExternalPlugins discovers the external plugins in the plugins root directory and adds them to external.Plugin.
+func DiscoverExternalPlugins() (ps []plugin.Plugin, err error) {
+	pluginsRoot := setPluginsRootBasedonHost()
+
+	rootInfo, err := os.Stat(pluginsRoot)
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			logrus.Debugf("External plugins dir %q does not exist, skipping external plugin parsing", pluginsRoot)
+			return nil, nil
+		}
+		return nil, err
+	}
+	if !rootInfo.IsDir() {
+		logrus.Debugf("External plugins path %q is not a directory, skipping external plugin parsing", pluginsRoot)
+		return nil, nil
+	}
+
+	infos, err := ioutil.ReadDir(pluginsRoot)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, info := range infos {
+		if !info.IsDir() {
+			logrus.Debugf("%q is not a directory so skipping parsing", info.Name())
+			continue
+		}
+
+		subinfos, err := ioutil.ReadDir(filepath.Join(pluginsRoot, info.Name()))
+		if err != nil {
+			return nil, err
+		}
+
+		for _, subinfo := range subinfos {
+			pluginInfos, err := ioutil.ReadDir(filepath.Join(pluginsRoot, info.Name(), subinfo.Name()))
+			if err != nil {
+				return nil, err
+			}
+			for _, pluginInfo := range pluginInfos {
+				// find the executable that matches the same name as info.Name().
+				// if no match is found, compare the external plugin string name before dot and match it with info.Name() which is the external plugin root dir.
+				// for example: sample.sh --> sample, externalplugin.py --> externalplugin
+				trimmedPluginName := strings.Split(pluginInfo.Name(), ".")
+				if len(trimmedPluginName) == 0 {
+					return nil, fmt.Errorf("Invalid plugin name found %q", pluginInfo.Name())
+				}
+
+				if pluginInfo.Name() == info.Name() || trimmedPluginName[0] == info.Name() {
+					// check whether the external plugin is an executable.
+					if !isPluginExectuable(pluginInfo.Mode()) {
+						return nil, fmt.Errorf("External plugin %s found in path is not an executable", pluginInfo.Name())
+					}
+
+					_, args := parseExternalPluginArgs()
+
+					ep := external.Plugin{
+						PName: info.Name(),
+						Path:  filepath.Join(pluginsRoot, info.Name(), subinfo.Name(), pluginInfo.Name()),
+						// todo(rashmigottipati): external.Plugin must satisfy the Plugin interface,
+						// so adding the supported project version as a placeholder value for now.
+						PSupportedProjectVersions: []config.Version{cfgv2.Version, cfgv3.Version},
+						Args:                      args,
+					}
+
+					if err := ep.PVersion.Parse(subinfo.Name()); err != nil {
+						return nil, err
+					}
+
+					logrus.Printf("Adding external plugin: %s", ep.Name())
+
+					ps = append(ps, ep)
+
+				}
+			}
+		}
+
+	}
+	return ps, nil
+}
```

Comment:
Some of the var names confused me. I read the entire method, I get the gist of what you are doing only because I conceptually know what you are trying to write. But the variables seemed weird to me. Specifically `infos`, `info`, `subinfos`, `subinfo`. 

Basically if `pluginsRoot` is `$XDG_CONFIG_HOME/kubebuilder/plugins/` then `infos, err := ioutil.ReadDir(pluginsRoot)` will return the list of plugin directories. So I think I'd use `pluginInfos` or `plugins` I don't care that they are `FileInfo` I care more what they represent. 

Then this line `subinfos, err := ioutil.ReadDir(filepath.Join(pluginsRoot, info.Name()))` will basically be returning the versions. So I'd say instead of `subinfos` I'd use `versions`. 

Then for `pluginInfos` I'd use `pluginFiles` or `pluginFileInfo`. This will tell me I'm at the level where the plugin should be defined.

I think this would make it easier to come back and fix bugs. Let's say there's a problem loading files from the plugins. Well you know it will be in the `pluginFiles` section. Or there's a problem reading the versions layer, you'd know it would be in the `versions` section.

Reply:
Agreed, it looks better this way and easier for debugging as well. Fixed it as per your recommendation.